File: problems_list/0031.下一个排列/0031-下一个排列.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size(), i=n-2;
        while(i>=0 && nums[i]>=nums[i+1]) i--;
        // 最后一个排列
        if(i>=0){
            for(int j=n-1; j>i; j--)
                if(nums[j]>nums[i]){
                    swap(nums[i], nums[j]);
                    break;
                }
        }
        reverse(nums.begin()+i+1, nums.end());
    }
};

File: problems_list/0031.下一个排列/0031-下一个排列.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    let n=nums.length, i=n-2;
    while(i>=0 && nums[i]>=nums[i+1]) i--;
    if(i>=0){
        for(let j=n-1; j>i; j--){
            if(nums[j]>nums[i]){
                let tmp=nums[i];
                nums[i]=nums[j];
                nums[j]=tmp;
                break;
            }
        }
    }
    for(let l=i+1, r=n-1; l<r; l++, r--){
        let tmp=nums[l];
        nums[l]=nums[r];
        nums[r]=tmp;
    }
};